src/backend/kms/surface/mod.rs

```diff
@@ -0,0 +1,1361 @@
+        let transform = output.current_transform();
+        let texture_buffer = TextureRenderBuffer::from_texture(
+            renderer,
+            texture,
+            1,
+            transform,
+            Some(opaque_regions),
+        );
+
+        let damage_tracker = OutputDamageTracker::from_output(output);
+
+        Ok(MirroringState {
+            texture: texture_buffer,
+            damage_tracker,
+        })
+    }
+}
+
+pub type GbmDrmCompositor = DrmCompositor<
+    GbmAllocator<DrmDeviceFd>,
+    GbmDevice<DrmDeviceFd>,
+    Option<(
+        OutputPresentationFeedback,
+        Receiver<(ScreencopyFrame, Vec<Rectangle<i32, BufferCoords>>)>,
+    )>,
+    DrmDeviceFd,
+>;
+
+#[derive(Debug)]
+pub enum QueueState {
+    Idle,
+    /// A redraw is queued.
+    Queued(RegistrationToken),
+    /// We submitted a frame to the KMS and waiting for it to be presented.
+    WaitingForVBlank {
+        redraw_needed: bool,
+    },
+    /// We did not submit anything to KMS and made a timer to fire at the estimated VBlank.
+    WaitingForEstimatedVBlank(RegistrationToken),
+    /// A redraw is queued on top of the above.
+    WaitingForEstimatedVBlankAndQueued {
+        estimated_vblank: RegistrationToken,
+        queued_render: RegistrationToken,
+    },
+}
+
+impl Default for QueueState {
+    fn default() -> Self {
+        QueueState::Idle
+    }
+}
+
+#[derive(Debug)]
+pub enum ThreadCommand {
+    Suspend,
+    Resume {
+        surface: DrmSurface,
+        gbm: GbmDevice<DrmDeviceFd>,
+        cursor_size: Size<u32, BufferCoords>,
+        vrr: bool,
+        result: SyncSender<Result<()>>,
+    },
+    NodeAdded {
+        node: DrmNode,
+        gbm: GbmAllocator<DrmDeviceFd>,
+        egl: EGLContext, // TODO: Option for software rendering
+    },
+    NodeRemoved {
+        node: DrmNode,
+    },
+    UpdateMirroring(Option<Output>),
+    VBlank(Option<DrmEventMetadata>),
+    ScheduleRender,
+    // TODO: Error handling
+    SetMode(Mode),
+    End,
+}
+
+#[derive(Debug)]
+pub enum SurfaceCommand {
+    SendFrames,
+    RenderStates(RenderElementStates),
+}
+
+impl Surface {
+    pub fn new(
+        output: &Output,
+        crtc: crtc::Handle,
+        connector: connector::Handle,
+        primary_node: DrmNode,
+        dev_node: DrmNode,
+        target_node: DrmNode,
+        evlh: &LoopHandle<'static, State>,
+        shell: Arc<RwLock<Shell>>,
+        startup_done: Arc<(Mutex<bool>, Condvar)>,
+    ) -> Result<Self> {
+        let frame_callback_seq = Arc::new(AtomicUsize::new(0));
+
+        let (tx, rx) = channel::<ThreadCommand>();
+        let (tx2, rx2) = channel::<SurfaceCommand>();
+        let active = Arc::new(AtomicBool::new(false));
+        let active_clone = active.clone();
+
+        let frame_callback_seq_clone = frame_callback_seq.clone();
+        let output_clone = output.clone();
+        let thread_handle = std::thread::spawn(move || {
```

Comment:
I think it would be a bit clearer to at least move the main function of the thread into it's own function, instead of having it all in the closure here.

Reply:
done